src/Compilers/CSharp/Portable/CodeGen/EmitArrayInitializer.cs

```diff
@@ -445,51 +492,154 @@ private bool TryEmitReadonlySpanAsBlobWrapper(NamedTypeSymbol spanType, BoundExp
+                // rather than `new ReadOnlySpan<char>(new[] { 'b', 'c' })`.  If such a pattern is found to be
+                // common, this could be augmented to accomodate it.  For now, we just return false to fail
+                // to optimize this case.
+                return false;
+            }
+
+            if (inPlaceTarget is not null)
+            {
+                // We can use RuntimeHelpers.CreateSpan / cached array but not for in-place initialization. Fail to optimize,
+                // but tell the caller they can call this again with a null inPlaceTarget, at which point this
+                // should be able to optimize the call.
+                avoidInPlace = true;
+                return false;
+            }
+
+            // As we're dealing with multi-byte types, endianness needs to be considered. Such handling is provided by the
+            // runtime's RuntimeHelpers.CreateSpan, which will wrap a span around the blob on little endian and which will
+            // allocate an array and cache it on big endian.
+            MethodSymbol? createSpan = (MethodSymbol?)_module.Compilation.GetWellKnownTypeMember(WellKnownMember.System_Runtime_CompilerServices_RuntimeHelpers__CreateSpanRuntimeFieldHandle);
+            if (createSpan is not null)
+            {
+                if (createSpan.GetUseSiteInfo().DiagnosticInfo?.Severity == DiagnosticSeverity.Error)
+                {
+                    // Something went wrong.
+                    return false;
+                }
+
+                // CreateSpan was available. Use it.
+                // ldtoken <PrivateImplementationDetails>...
+                // call ReadOnlySpan<elementType> RuntimeHelpers::CreateSpan<elementType>(fldHandle)
+                var field = _builder.module.GetFieldForData(data, alignment: (ushort)specialElementType.SizeInBytes(), wrappedExpression.Syntax, _diagnostics.DiagnosticBag!);
+                _builder.EmitOpCode(ILOpCode.Ldtoken);
+                _builder.EmitToken(field, wrappedExpression.Syntax, _diagnostics.DiagnosticBag!);
+                _builder.EmitOpCode(ILOpCode.Call, stackAdjustment: 0);
+                EmitSymbolToken(createSpan.Construct(arrayType.ElementType), wrappedExpression.Syntax, optArgList: null);
+                return true;
+            }
+
+            // We're dealing with a multi-byte primitive, and CreateSpan was not available.  Get a static field from PrivateImplementationDetails,
+            // and use it as a lazily-initialized cache for an array for this data:
+            //     new ReadOnlySpan<T>(PrivateImplementationDetails.ArrayField ??= RuntimeHelpers.InitializeArray(new int[Length], PrivateImplementationDetails.DataField));
+
+            var rosArrayCtor = (MethodSymbol?)Binder.GetWellKnownTypeMember(_module.Compilation, WellKnownMember.System_ReadOnlySpan_T__ctor_Array, _diagnostics, syntax: wrappedExpression.Syntax, isOptional: true);
+            if (rosArrayCtor is null)
```

Comment:
Should we do a similar `GetUseSiteInfo()` check for `WellKnownMember.System_ReadOnlySpan_T__ctor_Array` as above for `WellKnownMember.System_ReadOnlySpan_T__ctor_Array`?